### DISCUSSION TITLE
Accepting both JSON and form data in one route

### DISCUSSION THREAD

**Discussion by Terrance** [fastapi/fastapi#5666] [Questions] [answered]
I'm trying to implement an endpoint that can accept its request body either as JSON or as a multipart form (no files, just text fields), in order to satisfy an existing API specification.

As it stands, sending multipart to the endpoint with a Pydantic model just gives me a `422 Unprocessable Entity`.  The docs suggest I need to declare `fastapi.Form` parameters (there's no mention of models on the form pages as far as I can tell), which presumably means I can't use a model class for this?

So far the best I've come up with is a form data endpoint that rebuilds the model manually:

```python
class CreateData(BaseModel):
    first: str
    second: str
    ...

@router.post("", response_model=CreateResponse)
async def create_form(first: str = Form(), second: str = Form(), ...):
    data = CreateData.parse_obj({
        "first": first,
        "second": second,
        ...
    })
    return await create_json(data)

@router.post("", response_model=CreateResponse)
async def create_json(data: CreateData):
    ...
```

This appears to work, though it's clunky and I have to type all the fields out in multiple places.

I also need to accept both on the same HTTP method/path -- is there a way to overload a route and declare multiple handlers based on the incoming content type, or otherwise do this logic in a single handler?

I've also considered doing this transformation at the middleware level, though [it sounds like I shouldn't be trying to modify the body](https://github.com/tiangolo/fastapi/issues/5165).

**Comment by tiangolo** [comment]
It's probably better to ask in a GitHub issue, that will guide you to structure the question in a way that will be more easy to answer, create an example that I can copy paste and run, etc.

**Comment by Terrance** [comment, marked as answer]
After some more digging for potential workarounds, I eventually found #990 via Stack Overflow, so I guess this is a won't-fix.

Instead, I made a small dispatching dependency to handle Pydantic model creation dynamically:

```python
_TModel = TypeVar("_TModel", bound=BaseModel)

def form_or_json(model: Type[_TModel]) -> _TModel:
    async def form_or_json_inner(request: Request) -> _TModel:
        type_ = request.headers["Content-Type"].split(";", 1)[0]
        if type_ == "application/json":
            data = await request.json()
        elif type_ == "application/x-www-form-urlencoded":
            data = await request.form()
        else:
            raise HTTPException(400)
        return model.parse_obj(data)
    return Depends(form_or_json_inner)
```

The result is a handler that will accept both JSON and form request bodies, whilst declaring the model name and fields only once:

```python
@router.post("", response_model=CreateResponse)
async def create(data = form_or_json(CreateData)):
    ...
```
